csharp/src/Google.Protobuf.Test/Reflection/DescriptorDeclarationTest.cs

```diff
@@ -0,0 +1,151 @@
+﻿#region Copyright notice and license
+// Protocol Buffers - Google's data interchange format
+// Copyright 2018 Google Inc.  All rights reserved.
+// https://developers.google.com/protocol-buffers/
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+using Google.Protobuf.Reflection;
+using NUnit.Framework;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Protobuf.Test.Reflection
+{
+    // In reality this isn't a test for DescriptorDeclaration so much as the way they're loaded.
+    public class DescriptorDeclarationTest
+    {
+        static readonly FileDescriptor unitTestProto3Descriptor = LoadProtos();
+
+        [Test]
```

Comment:
Test FileDesciptor comments as well?

Reply:
Ah, I didn't think FileDescriptors could have comments themselves. Will add that support in another commit, and let you know when I've done so.

Reply:
Might  unable to get the leading comments which are in "an unstable location" mentioned in this issue:
https://buganizer.corp.google.com/issues/18735548

We may only able to add comments for FileDescriptor's package, syntax etc...

Reply:
Given the comments on that bug, I'd suggest leaving `FileDescriptor` without a `Declaration` property at all for the moment. It will be easy enough to add later if we want to. As such, I think the PR should be okay in its current form. Let me know if you disagree though - I'd be happy to be convinced :)

Reply:
Agree, just add a comments for the FileDescriptor issue.

Reply:
I've added two commits - one with a comment for FileDescriptor, and one tidying up the other descriptor-based tests.